server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -574,6 +575,23 @@ public Settings getSettings() {
         return this.aliases;
     }
 
+    /**
+     * Lazy loaded cache for tier preference setting. We can't eager load this setting because
+     * {@link IndexMetadataVerifier#convertSharedCacheTierPreference(IndexMetadata)} might not have acted on this index yet and thus the
+     * setting validation for this setting could fail for metadata loaded from a snapshot or disk after an upgrade.
+     */
+    private List<String> tierPreference;
```

Comment:
Suggest adding to the comment that this needs no synchronisation since it's a pure function of the immutable `settings` object, similarly to how `String#hash` works. Also this:

```suggestion
    @Nullable // since lazy-loaded
    private List<String> tierPreference;
```

Reply:
++ added